Prevent Security Group Shopping

eschultz17 opened the pull request:
- [x] Modify `network-location` filter:
   * Add ability to ignore based on tag value
   * Add new network-location reason: `SecurityGroupMismatch`

- [x] Modify `modify-security-groups` action:
  * add `remove` option: `network-location`
      * Removes security groups that do not match the given key/tag
      * Searches `c7n:NetworkLocation` array for the `SecurityGroupMismatch` object and removes those security groups (ignores other objects/reasons in `c7n:NetworkLocation` array)

Example policy:
```
  - name: ec2-mismatched-sg-remove
    resource: ec2
    filters:
      - type: network-location
        compare: ["resource","security-group"]
        key: "tag:TEAM_NAME"
        ignore:
          - "tag:TEAM_NAME": ENTERPRISE
    actions:
      - type: modify-security-groups
        remove: network-location
        isolation-group: sg-xxxxxxxx
```

Example `c7n:NetworkLocation` key:
```
    "c7n:NetworkLocation": [
      {
        "reason": "SecurityGroupLocationCardinality", 
        "security-groups": {
          "sg-xxxxxxx1": "SUPPORT", 
          "sg-xxxxxxx2": "DEVELOPMENT",
          "sg-xxxxxxx3": "DEVELOPMENT",
        }
      }, 
      {
        "security-groups": {
          "sg-xxxxxxx1": "SUPPORT", 
        }, 
        "reason": "SecurityGroupMismatch", 
        "resource": "DEVELOPMENT"
      }
    ], 
```

kapilt commented:
re network-location filter doc updates, the example policy you have in the pr would be a good fit.

eschultz17 commented:
@kapilt I added the example to the network-location documentation. Let me know if you'd like me to add more information for either the modify-security-groups actions or network-location filter docs. 

I removed the `jmespath.search` from the network-location filter since `get_resource_value` appears to yield the same results. 

also, left a comment here:https://github.com/capitalone/cloud-custodian/pull/2670#discussion-diff-205791298R231 explaining why I left the SecurityGroupMismatch results as a dictionary vs an array

kapilt commented on c7n/actions.py:
```diff
@@ -276,6 +277,10 @@ def get_groups(self, resources, metadata_key=None):
+            elif remove_target_group_ids == 'network-location':
+                for reason in r.get('c7n:NetworkLocation', ()):
+                    if reason['reason'] == 'SecurityGroupMismatch':
+                        remove_groups = [sg for sg in reason['security-groups']]
```
the filter implementation is using a dict for the annotation now,  so this usage is invalid now, given that it seems like this functionality should have a unit test.

kapilt commented on c7n/filters/vpc.py:
```diff
@@ -221,6 +237,15 @@ def process_resource(self, r, resource_sgs, resource_subnets, key):
                     'resource': r_value,
                     'subnet': subnet_values})
+            if 'security-group' in self.compare and resource_sgs:
+                temp_sgs = {sg_id: sg_value
```
preference is for an actual name, mismatched_sgs, temp is too generic.

kapilt commented on c7n/filters/vpc.py:
```diff
@@ -221,6 +237,15 @@ def process_resource(self, r, resource_sgs, resource_subnets, key):
+                temp_sgs = {sg_id: sg_value
+                            for sg_id, sg_value in sg_values.items()
+                            if sg_value != r_value}
+                if len(temp_sgs) > 0:
```
just use an `if mapping`, the length check is a bit superflous given natural boolean evaluation of dicts.
